### ISSUE TITLE
Enhancement idea: Localizations as a cloze (German: Lückentext)

### ISSUE THREAD

**Issue by jongleur1983** [SeriousM/WPFLocalizationExtension#78]
Once in a while I come across a UI pattern that includes controls in a sentence.
Like 
"The shop is open [input1] from [input2] to [input3].",
where input1 is a control to select some days of the week, and input2 and input3 are inputs for the time of the day.

For localization it would be great if it could be handled with a localization format string like 
"The shop is open {0} from {1} to {2}", where the localization handles inserting controls like it would insert text parameters in usual format-strings.

I think about implementing a mechanism like this and enhancing the WPFLocalizationExtension might be one place for it to live.

I see that the WPFLocalizationExtension isn't actively developed currently, but would it be welcome to add such a feature at all?

**Comment by SeriousM** [comment]
Hi!

The reason why it looks not active is because it's pretty stable, not dead
:)

I've seen the need for such a feature over the years over and over again,
but never had a good idea to implement it so that it's really useful.

It would be awesome to soo your PR for that! If you need some help to the
api don't hesitate to ask, we'll try to answer your questions.

Cheers
Am 31.10.2015 10:27 schrieb "jongleur1983" <notifications@github.com>:

**Comment by jongleur1983** [comment]
working on it right now I would ask for some more ideas, what should be supported by a feature like this.

I propose something like the following xaml code for usage:

```xaml
<loc:Cloze FormatString="The shop in {0} is open on {0} from {1} to {2}"
            DataContext="{Binding ContainerElement}">
    <GapText.Gaps>
        <system:String>{Binding City}</system:String>
        <ComboBox x:Name="DayOfTheWeek" ItemsSource="{Binding WeekDays}" />
        <TextBox x:Name="openingTime" Text="{Binding OpeningTime}" />
        <TextBox x:Name="closingTime" Text={Binding ClosingTime}" />
    </GapText.Gaps>
</loc:Cloze>
```

Applied this would render to something like this (where all children of the outer TextBlock would of course be the result of the bound FormatString and Gaps):

```xaml
<TextBlock HorizontalAlignment="Stretch" 
           TextWrapping="Wrap"
           TextDecorations="Underline"
           LineStackingStrategy="BlockLineHeight"
           DataContext="{Binding OpenDayItem}">
    <Run Text="The shop in " />
    <Run Text="{Binding City}" />
    <Run Text=" is open on " />
    <ComboBox x:Name="DayOfTheWeek" ItemsSource="{Binding WeekDays}" />
    <Run Text=" from " />
    <TextBox x:Name="openingTime" Text="{Binding OpeningTime}" />
    <Run Text=" to " />
    <TextBox x:Name="closingTime" Text="{Binding ClosingTime}" />
    <Run Text="." />
</TextBlock>
```
Where City binds to OpenDayItem.City and so on.

I would like to hear alternative ideas on how the elements may be named as well. As a German native speaker Cloze was not known to me until I searched for the right English term, so I don't know if it's the most common one. Probably there is something else that could be used?

**Comment by SeriousM** [comment]
As you write it here, the FormatString should be localizable, right?
How do you like to handle missing elements (2 placeholders in text but 1
gap text provided) or duplicate elements (eg twice {0} in the string)?
Cloze is also new to me a d I never heard of it till now. What about
loc:segment(ed) with segment(ed).part?
Am 01.12.2015 07:54 schrieb "jongleur1983" <notifications@github.com>:

**Comment by jongleur1983** [comment]
I would not handle missing elements at all. If that happens, the corresponding element is not included in the visual tree afterwards, but that's straight in line with how localization with text placeholders work.

In fact, for usual pure-text localization a localized string value that lacks any of the placeholders is a usage error as well - with one single exception: if the same fact is encoded in several different ways as different parameters. But this exception of the rule - and it's the only one I can imagine yet - is the same for controls:
What if the software developer provides two different ways to set and get e.g. date and time? there might be locales that are more used to provide a date by a calendar, while others use something different. Thus there may be two (or even more) controls that bind to the same value and not necessary to use in the UI at all.

So if you consider the possibility of unused, but provided placeholders a problem, you should consequently see the same possibility a problem for the text localization as a "wrong" string value omits some information the developer might have considered necessary or essential to be provided at that place.

**Comment by SeriousM** [comment]
Well, that's true, but as a developer, or have to consider that the user of
the software will make mistakes and you need a way to tell him/her that.
That's the reason why we show the localization key instead of N empty
string for not-found elements.
Anyway, I guess you want to support binding to the segments, right? How do
you want to handle language switching (a control needs to be repositioned
but keep the value)?
Cheers
Am 01.12.2015 23:17 schrieb "jongleur1983" <notifications@github.com>:

**Comment by jongleur1983** [comment]
Regarding the fallback to display the localization key: agree - but if there is a localization key up to know where two parameters are provided in the loc-binding, but only one occurs in the localization string, there's nothing to indicate that yet, right? so a ```{lex:LocText MyKey FormatSegment1={Binding SomeStringProperty} FormatSegment2={Binding SomeIntProperty} FormatSegment3={Binding SomeDateProperty}}``` can be used and if the value behind MyKey is "This is a string with string value {0} and date value {2}" the library wouldn't say anything about it - how should it at all, as that string might not even be known at compile time.

About keeping the binding:
I have to confess: I'm not entirely sure if it's going to work the way I imagine it, but as far as I know Binding usually works in the logical tree of WPF, while all I have to manipulate here is the visual tree: Reorder of the visual children, and probably initially inserting a visual element between the GapText control element and the controls provided for the gaps.

**Comment by GeroIwan** [comment]
Is this really a WPFLocalizationExtension issue? The Cloze control itself is independent from localization. Only its FormatString property would have to get its value via WPFLocalizationExtension. But providing localized property values is usual business.

**Comment by jongleur1983** [comment]
@GeroIwan Is that use case "usual business"? I would be happy if you could share your knowledge on how you would solve the issue.
Can you give an example how you do that as usual business?

Such an example should be able to
* include two or more placeholders that are not displayed as text
* include the possibility to switch their order depending on the locale

As an example, when locale 1 displays ```string control1 string control2 string```, it should be simple to localize that to locale2 as ```string control1 control2 string``` or to locale3 as ```control2 string control1```.

I'm happy to hear that this is "usual business", but I think it's a fairly hard task - a task that's not used very often, but much work to do for the rare cases of a single project where it would be useful.

**Comment by GeroIwan** [comment]
@jongleur1983 What I wanted to say is:
* It is _usual localization business_ to have a resource key whose value in locale1 is `"lorem {0} ipsum {1} dolor"`, in locale2 is `"sit {0} {1} amet"`, and in locale3 is `"{1} consectetur {0}"`, and to provide the value to a string property of some control (e.g. a FormatString property).
* The other part of the issue is having a control (say, `Cloze`) that takes a FormatString (which, of course, may change (whether by binding or by localization does not matter - it simply changes)) plus some controls, and that replaces the placeholders within the FormatString by the controls. However, this part is _no localization problem_. (But yes, this may be a fairly hard task. Unfortunately, I do not have time to investigate it at the moment.)
* Conclusion: Localization is only involved when providing the FormatString (= usual localization business), the actual control replacement then is a task on its own (= no localization problem).
* Hence, again: Is this really a WPFLocalizationExtension issue? I don't think so. (But I think that the `Cloze` control, on its own, is worth a development.)

Do you agree? Or am I missing something?

**Comment by jongleur1983** [comment]
@GeroIwan 
Agree up to your conclusion.
When you look into the source code of the project there are two controls already. 
Maybe those are seldom in use, I discovered them when I investigated if it's useful to have this included in WPFLocalizationExtension or not (the same question you raise here).

These are at least:

* EnumRun is a Run derivate that binds to an Enum value, see  https://github.com/SeriousM/WPFLocalizationExtension/blob/master/WPFLocalizeExtension/Engine/EnumRun.cs 
* EnumComboBox derives from ComboBox to bind to a whole Enum https://github.com/SeriousM/WPFLocalizationExtension/blob/master/WPFLocalizeExtension/Engine/EnumComboBox.cs

Yes, those are helpers. It's more easy to have a combobox that allows to bind to an enum and supports translating the enum-Value to a corresponding localized, human-readable string. Not more but not less.

So the Cloze would be similar or the same.

But in fact: if I get it to work and it's not wanted in this project I agree that it's worth the effort nevertheless and I would publish it in an own repository.

**Comment by SeriousM** [comment]
I would welcome such a control as long as it offers a range of features
instead just cover a single case ;)

**Comment by jongleur1983** [comment]
I'm working on it.
May take a while though. Feel free to provide necessary or nice-to-have features.

**Comment by SeriousM** [comment]
Just keep us informed and we can give you our thoughts to it.

**Comment by GeroIwan** [comment]
@jongleur1983  Just as addendum:
* EnumRun is really a _localization control_: It uses the bound Enum value to compute the key for a private LocExtension member that in turn is "bound"(assigned/provided) to the Text property of the underlying Run.
* EnumComboBox in fact does not involve localization, and hence is a helper that could be moved to a different project. In the context of localization, it's probably intended to be used in conjunction with TextBlock+EnumRun within the ItemTemplate for the ComboBox.

@SeriousM [a bit off topic] Looking at the last point, why not include the ItemTemplate within EnumComboBox initially (= default; ItemTemplate can still be overwritten (e.g. in XAML).) in order to make it a _localization control_ out of the box since it is part of WPFLocalizationExtension? (Of course, EnumRun properties (except EnumValue) would have to be passed/bound through.)

**Comment by SeriousM** [comment]
The controls are includes by @MrCircuit and I don't know anything about
them, but your guess sounds good!

**Comment by jongleur1983** [comment]
First Update, although I didn't upload anything yet.
Sketch of the actual approach:

- Custom Control (planned to support templates). Currently the outermost control is a TextBlock as I want to support the inline-elements variant first (where text can wrap anywhere like it's done in a TextBlock).
- A Property with name `FormatString` takes a formatString, this will usually be used together with the WPFLocalizationExtension to provide the actual localization.
- A second property, currently named `Gaps` takes the gaps themselfes. Gaps are added like Items to an ItemsControl. These Gaps are handled as follows: 
 - Any Inline object is used as is in the items. 
 - Any UIElement is used with it's implicit Run-Container
 - Any other item is handled like in the usual string.Format binding, allowing any format specifier etc.

- The FormatString is splitted using the regex `@"(.*?){(\d*)}"`. the latter part `{(\d*)}?` captures each "simple" format parameter, where "simple" means there's no additional specifier for padding, format or similar. For now I'm going to support only "simple" format parameters to be replaced by controls (or more specifically: UIElements). A successful match of this regex contains the following groups
 - [0] is the complete group, that consists of some string followed by a single simple format parameter. My example from beginning of the thread would split into the parts `The shop in {0}`, ` is open on {1}`, ` from {2}` and ` to {3}`.
 - [1] is the string before the placeholder, for the first match of the same example, this would be `The shop in `
 - [2] is the number of the placeholder, here just `0`.

Let's consider the same example but parametrize the weekday to be a corresponding format parameter on a DateTime binding, and add some suffix as that's another corner case:
`The shop in {0}  is open on {1:dddd} from {2} to {3} every week` 

As specified in https://msdn.microsoft.com/de-de/library/8kb3ddd4%28v=vs.110%29.aspx `dddd` formats a `DateTime` object as full day of the week name according to current culture.

The regex splits this into the matches
- `The shop in {0}`
- ` is open on {1:dddd} from {2}` and
- ` to {3}` 
...where the groups are as before, so we can take for the `Inline` property of the TextBlock:
- the string `The shop in ` (group [1] of first match)
- Gaps[0], where 0 is group[2] of first match
- the string ` is open on {1:dddd} from ` group[1] on second match
- Gaps[2], where 2 is group[2] on second match
- the string ` to `, group[1] on third match
- Gaps[3], where 3 is group[2] on third match.

Now we miss the suffix, but from the last match we know where it ends, so we can take the suffix as a simple substring and add it as the last element.
Of course the string items that come from group[1] of the matches have to be formatted themselfes. We format these by the whole gaps property as arguments. Thus any parameter that is not simple could in theory take any parameter, even the UIElements or inlines, but in that case those are for now replaced by their toString() equivalent.

# To sum it up:
With this approach I think I could reach the following:
- Enable using UIElements in a format string to show them as themselves
- Keep all features for bound datatypes where `string.Format()` is useful already, including Format specifications

# Possible Enhancements
- I'm not sure wether the restriction to "simple" parameters is useful. Some UIElements used here may use the string for display purpose only, so here propagating even format specifiers may be useful.
Up to now I don't have a clue how to solve that elegantly, so it's only an idea without any guarantee to be supported.

**Comment by jongleur1983** [comment]
update: I don't get the default template for the control to work properly, but with a custom template it seems to work. Some cleanup is missing yet and some more tests to apply, but I'm on a good way, I think. Hope to be able to commit something for review before christmas at least.

**Comment by jongleur1983** [comment]
feel free to take a look into the commit on my branch https://github.com/jongleur1983/WPFLocalizationExtension/tree/feature/GapText - it contains a small demo project WPFLocalizationExtensionDemoApplication which I would like to extend to a general demo project for the whole WPFLocalization. Up to know it contains demos for the GapTextControl.

Currently the limitations I would see as "bugs" are:
* does not work with direct binding for the Gaps. Has to use RelativeSource binding with ancestor anchor at the GapTextControl or above.
* tests done yet are very limited.

I would be happy to get your opinions, feel free to comment, although I'll not be available from somewhen tomorrow until approx January 3rd.
Thus: Merry Christmas and a happy new year, and thanks so far.

**Comment by SeriousM** [comment]
Thank you very much for the work! I'll have a look next year because it's
family time.
Frohe Weihnachten und nen guten Rutsch!

**Comment by MrCircuit** [comment]
Hi there,

thanks for your input. It looks promising but I encountered an issue that will definitely restrict the usage: Due to the WPF concept of single-parent-relationship (a child can not have many parents) a format string like this will not work with the current approach: "The shop in {0} is great. We need more of its kind in {0}." (OK, this text sounds weird - but think of the issue in general :-))

This would try to place the Inline or UIElement of {0} twice in the topmost TextBlock of the control - which will not work. Any idea how to solve this?

@GeroIwan [BTW]: I changed the EnumComboBox as you proposed. Its Items are now by default templated with the EnumRun component.

**Comment by jongleur1983** [comment]
Thanks for the feedback.
I agree that the case you mentioned has to be handled.

The following code throws an error as well and is independent of how my extension works. I guess (without having tested it) it would throw an exception when used with the current WPFLocalizationExtension as well:

```
using System;
					
public class Program
{
	public static void Main()
	{
		decimal price = 169.32m;
      	Console.WriteLine("The cost is {0:Q2}.", price);
	}
}
```

This throws a System.FormatException: Format specifier was invalid.

It's a runtime exception and is not found at compile time (although resharper might help to find some of those cases).

Similarly my code could be enhanced to throw a FormatException as well, something like "The format string 'The shop in {0} is great. We need more of it's kind in {0}.' tries to use the same FrameworkElement more than once.".

This cannot prevent all errors of this kind, but it's not different from what happens now.
Preventing all of these errors is impossible when we think on the different sources the localization strings can come from by different provider implementations, thus they are not always known at compile time.

**Comment by MrCircuit** [comment]
I found a solution to this issue. I introduced a simple deep copy method using XamlWriter/Reader and Binding expression copy. It works but I would be glad if you could double check this.

We should think about:
* Make it faster - we could at least store the XAML string of each object of the gap collection thus reducing the XamlWriter effort
* Memory leaks - is it safe to replace already cloned objects by new ones (e.g. happens often at the beginning, when the Gaps and further properties are set one by one)?

**Comment by jongleur1983** [comment]
My major concern against that solution proposal is that a pure copy might not be enough.

Let's take an element that has an ```x:Name``` and is referred to by this name from somewhere else.
If I'm not mistaken this is possible even when the element is not necessarily part of the visual tree, so we can have those references to FrameworkElements that are items.

In this case we would with your proposal copy an element to two and either loose one x:Name property or violate the constraint that x:Name must be unique inside an XAML name scope (see https://msdn.microsoft.com/en-us/library/ms752290%28v=vs.110%29.aspx for reference).

Violating that constraint isn't a good idea obviously, but losing the x:Name for one (arbitrary?) copy of the element isn't either, as the functionality get's hard to predict for the developer, and the expectation that the element - if bound twice - behaves the same - is not met.

As we cannot determine easily when that works and when it doesn't it would have to be highly discouraged to use the same FrameworkElement reference twice. Instead the best practice would always be to bind identical elements more than once to the items list if needed, as it improves maintainability (replace one element by something else) and predictability for developers and probably even users.

Therefore I propose to not allow duplicate item binding for frameworkElement gaps with the exception thrown as proposed before. Best practice if necessary would be to use elements with a common Template or style to reduce duplicate code.

Feel free to give counter arguments ;)

Nevertheless I'm not completely satisfied with the underlying issue that it's indeed a source of failure that is - without an additional idea - not possible to determine at compile time.

**Comment by MrCircuit** [comment]
The values of the Name and x:Name fields seem to have no impact on the runtime. Nevertheless I understand your point - there could be rare cases where these fields are needed. Appart from this, event registrations also fail to copy (a proper MVVM design may also get along without them).

In my opinion I would still stick to the current solution.

Another possible approach: Rather than building a deep copy we could also think about an instance that is Two-Way-bound to every single DependencyProperty of the original item. This would at least create a similar (still not all events may fire) behaviour compared to a single instance.

**Comment by jongleur1983** [comment]
@MrCircuit using the current version where you implemented the deepCopy variant doesn't work for me.
Starting the DemoApplication I get a XamlParseException at the XamlReader.Load() command in DeepCopy.
Just starting the WPFLocalizationExtensionDemoApplication is enough to throw the exception.

Exception details:

> System.Windows.Markup.XamlParseException ist aufgetreten.
>   HResult=-2146233087
>   LineNumber=1
>   LinePosition=501
>   Message=Zeilennummer "1" und Zeilenposition "501" von "Durch Hinzufügen eines Werts zu einer Auflistung vom Typ "System.Windows.Controls.ItemCollection" wurde eine Ausnahme ausgelöst.".
>   Source=PresentationFramework
>   StackTrace:
>        bei System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
>        bei System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, Boolean skipJournaledProperties, Uri baseUri)
>        bei System.Windows.Markup.XamlReader.Load(XamlReader xamlReader, ParserContext parserContext)
>        bei System.Windows.Markup.XamlReader.Load(XmlReader reader, ParserContext parserContext, XamlParseMode parseMode)
>        bei System.Windows.Markup.XamlReader.Load(XmlReader reader)
>        bei WPFLocalizeExtension.Engine.GapTextControl.DeepCopy[T](T obj) in N:\Users\jongleur\Documents\Git-Projekte\WPFLocalizationExtension\WPFLocalizeExtension\Engine\GapTextControl.cs:Zeile 183.
>   InnerException: 
>        HResult=-2146233079
>        Message=Der Vorgang ist während der Verwendung von "ItemsSource" ungültig. Verwenden Sie stattdessen "ItemsControl.ItemsSource", um auf Elemente zuzugreifen und diese zu ändern.
>        Source=PresentationFramework
>        StackTrace:
>             bei System.Windows.Controls.ItemCollection.CheckIsUsingInnerView()
>             bei System.Windows.Controls.ItemCollection.Add(Object newItem)
>             bei System.Xaml.Schema.XamlTypeInvoker.AddToCollection(Object instance, Object item)
>             bei MS.Internal.Xaml.Runtime.ClrObjectRuntime.Add(Object collection, XamlType collectionType, Object value, XamlType valueXamlType)
>        InnerException: 

I don't know what's wrong there, but it seem's not to be a valid solution so far ;)

**Comment by jongleur1983** [comment]
I have to admi that the exception I posted above occurs in a different test project I started a few weeks ago but didn't publish (yet). Investigating the problem furhter currently.

**Comment by jongleur1983** [comment]
still no idea why that fails, but tried a little bit more, compare this commit on my forks branch: https://github.com/jongleur1983/WPFLocalizationExtension/commit/ab110bff17eee7b4e6c9c93e4ed0ed8495838496

In Line 42 of GapTextWpfExample.xaml the EnumComboBox is commented out and replaced by an ordinary ComboBox for now. Using the EnumComboBox instead leads to the error posted above. The EnumComboBox in the upper part of that view (xaml line 26) works as expected, so I don't think it's something with how I use the EnumComboBox as such, but has to do with the GapText control or the DeepCopy mechanism proposed/implemented by @MrCircuit. Therefore, if you have any idea: I have no clue why/what's going wrong here. Happy to get more insight, but XamlReader/Writer and that stuff is new to me.

Thanks.

**Comment by MrCircuit** [comment]
Found a solution to it by fixing the serialization behaviour of the EnumComboBox...

**Comment by jongleur1983** [comment]
nice, thanks.
Only had a quick glance at what you do there; but it raises the question if the solution is stable.
What's the (general) requirement for the solution to work? 
Correct serialization behaviour?

If we need a hack for the EnumComboBox, how much special treatment is required for arbitrary controls?

**Comment by MrCircuit** [comment]
You are absolutely right - the solution fixes the EnumComboBox (btw: it does not cover inline item templates with binding...) and others might need adopted hacks too. I suggest to introduce an attached property for GapTextControl that lets the user decide if the particular control should copy or not.

**Comment by SeriousM** [comment]
Closed due stall conversation
